[scroll-animations] Add Animation.rangeStart and Animation.rangeEnd API

nmoucht opened the pull request:
#### 6950f6db7b42ae69add723c4656a556e8b317357
<pre>
[scroll-animations] Add Animation.rangeStart and Animation.rangeEnd API
<a href="https://bugs.webkit.org/show_bug.cgi?id=280300">https://bugs.webkit.org/show_bug.cgi?id=280300</a>
<a href="https://rdar.apple.com/136626109">rdar://136626109</a>

Reviewed by Antoine Quint.

Add Animation.rangeStart and Animation.rangeEnd API.

* Source/WebCore/animation/WebAnimation.cpp:
(WebCore::WebAnimation::setBindingsRangeStart):
(WebCore::WebAnimation::setBindingsRangeEnd):
* Source/WebCore/animation/WebAnimation.h:
(WebCore::WebAnimation::bindingsRangeStart const):
(WebCore::WebAnimation::bindingsRangeEnd const):
* Source/WebCore/animation/WebAnimation.idl:

Canonical link: <a href="https://commits.webkit.org/284367@main">https://commits.webkit.org/284367@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/56d7b9ea3ea216df58ac9fe3404df077dd8cdd49

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69005 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48405 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21677 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/73085 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/20173 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/71122 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/56206 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20011 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54954 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13409 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72071 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44191 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59575 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35428 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40856 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/17004 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18537 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62805 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17350 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74794 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12987 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16598 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/62597 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/13025 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59658 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62494 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10475 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4083 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10564 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/44209 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45282 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46478 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45024 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

graouts commented on Source/WebCore/animation/TimelineRangeOffset.h:
```diff
@@ -0,0 +1,40 @@
+#pragma once
+
+#include "CSSNumericValue.h"
+
```
No need for this blank line.

nmoucht replied:
Removed.

graouts commented on Source/WebCore/animation/TimelineRangeOffset.idl:
```diff
@@ -0,0 +1,32 @@
+    JSGenerateToNativeObject,
+] dictionary TimelineRangeOffset {
+  DOMString rangeName;
+  CSSNumericValue offset;
```
This needs to use 4-space indentation.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/TimelineRangeOffset.idl:
```diff
@@ -0,0 +1,32 @@
+[
+    JSGenerateToJSObject,
+    JSGenerateToNativeObject,
+] dictionary TimelineRangeOffset {
```
Assuming this is supported here, this should be guarded by `EnabledBySetting=ScrollDrivenAnimationsEnabled` I believe.

nmoucht replied:
Added.

graouts commented on Source/WebCore/animation/WebAnimation.idl:
```diff
@@ -54,6 +54,8 @@ typedef (double or CSSNumericValue) CSSNumberish;
     [ImplementedAs=bindingsStartTime] attribute CSSNumberish? startTime;
     [ImplementedAs=bindingsCurrentTime] attribute CSSNumberish? currentTime;
+    [ImplementedAs=bindingsRangeStart, EnabledBySetting=ScrollDrivenAnimationsEnabled] attribute (TimelineRangeOffset or CSSNumericValue or CSSKeywordValue or DOMString) rangeStart;
+    [ImplementedAs=bindingsRangeEnd, EnabledBySetting=ScrollDrivenAnimationsEnabled] attribute (TimelineRangeOffset or CSSNumericValue or CSSKeywordValue or DOMString) rangeEnd;
```
No need for `ImplementedAs` here since we don't need to do anything specific for when the values are set through the bindings. This means that the getter and setter on `WebAnimation` should remove the `Bindings` bit in their name.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/WebAnimation.h:
```diff
@@ -237,6 +244,9 @@ class WebAnimation : public RefCounted<WebAnimation>, public EventTarget, public
     ReplaceState m_replaceState { ReplaceState::Active };
     uint64_t m_globalPosition { 0 };
+    std::variant<TimelineRangeOffset, RefPtr<CSSNumericValue>, RefPtr<CSSKeywordValue>, String> m_rangeStart { };
+    std::variant<TimelineRangeOffset, RefPtr<CSSNumericValue>, RefPtr<CSSKeywordValue>, String> m_rangeEnd { };
```
It's not clear to me what `{ }` initializes this value to by default.

nmoucht replied:
Changed this to initialize to normal.

graouts commented on Source/WebCore/animation/KeyframeAnimationOptions.h:
```diff
@@ -28,14 +28,18 @@
     String id;
     std::optional<RefPtr<AnimationTimeline>> timeline;
     std::variant<FramesPerSecond, AnimationFrameRatePreset> frameRate;
+    std::variant<TimelineRangeOffset, RefPtr<CSSNumericValue>, RefPtr<CSSKeywordValue>, String> rangeStart;
```
`std::variant<TimelineRangeOffset, RefPtr<CSSNumericValue>, RefPtr<CSSKeywordValue>, String>` is used in quite a few places. Might be worth `typedef`'ing it in `WebAnimationTypes.h` for convenience, maybe as `TimelineRangeValue` or something of that ilk?

nmoucht replied:
Fixed.

nmoucht committed:
[scroll-animations] Add Animation.rangeStart and Animation.rangeEnd API
https://bugs.webkit.org/show_bug.cgi?id=280300
rdar://136626109

Reviewed by Antoine Quint.

Add Animation.rangeStart and Animation.rangeEnd API.

* Source/WebCore/animation/WebAnimation.cpp:
(WebCore::WebAnimation::setBindingsRangeStart):
(WebCore::WebAnimation::setBindingsRangeEnd):
* Source/WebCore/animation/WebAnimation.h:
(WebCore::WebAnimation::bindingsRangeStart const):
(WebCore::WebAnimation::bindingsRangeEnd const):
* Source/WebCore/animation/WebAnimation.idl:

Canonical link: https://commits.webkit.org/284367@main
